src/main/java/com/intuit/graphql/orchestrator/batch/QueryOptimizer.java

```diff
@@ -6,99 +6,68 @@
-                else
-                    selectionSetMap.put(f.getName(), new HashSet<>(getSelectionSetFields(f)));
-                createSelectionSetTree(f.getSelectionSet(), selectionSetMap);
-            }
-        });
-    }
+  public static ImmutablePair<List<Selection>, Map<String, List<Field>>> groupSelections(List<Selection> selections) {
+    List<Selection> distinctSelections = new ArrayList<>();
+    Map<String, List<Field>> fieldMap = new HashMap<>();
+    selections.forEach(selection -> {
+      //group fields with same name
+      if (selection instanceof Field) {
+        Field field = (Field) selection;
+        fieldMap.computeIfAbsent(field.getName(), k -> new ArrayList<>()).add(field);
+        // collect non fields.
+      } else {
+        distinctSelections.add(selection);
+      }
+    });
+    return new ImmutablePair<>(distinctSelections, fieldMap);
+  }
 
-    public static List<Field> getSelectionSetFields(Field f){
-        return f.getSelectionSet().getSelections()
-                .stream()
-                .map(s -> (Field) s)
-                .collect(Collectors.toList());
+  public SelectionSet getTransformedSelectionSet(SelectionSet selectionSet) {
+    if (operationType == Operation.QUERY) {
+      return recurse(selectionSet);
     }
+    return selectionSet;
+  }
 
-    public static SelectionSet mergeFilteredSelection(Field node, HashMap<String, Set<Field>> selectionSetTree) {
-        SelectionSet.Builder selectionSetBuilder = SelectionSet.newSelectionSet();
-        SelectionSet.Builder childrenSelectionSetBuilder = SelectionSet.newSelectionSet();
-        if (node.getSelectionSet() == null) { // check if leaf node
-            selectionSetBuilder.selection(node);
-            return selectionSetBuilder.build();
-        }
-        List<SelectionSet> childrenSelectionSets = new ArrayList<>();
-        for (Selection selection : selectionSetTree.get(node.getName())) {
-            childrenSelectionSets.add(mergeFilteredSelection((Field) selection, selectionSetTree));
-        }
-        childrenSelectionSets
-                .stream()
-                .map(s -> s.getSelections())
-                .flatMap(Collection::stream)
-                .forEach(childrenSelectionSetBuilder::selection);
-        selectionSetBuilder.selection(node.transform(builder ->
-                builder.selectionSet(childrenSelectionSetBuilder.build())));
+  public SelectionSet recurse(SelectionSet selections) {
+    SelectionSet.Builder mergedSelectionSetBuilder = SelectionSet.newSelectionSet();
+    final ImmutablePair<List<Selection>, Map<String, List<Field>>> selectionSetTree = groupSelections(
+        selections.getSelections());
+    selectionSetTree.getLeft().forEach(selection -> mergedSelectionSetBuilder.selection(selection));
+    selectionSetTree.getRight().values().forEach(fields -> mergedSelectionSetBuilder.selection(mergeFields(fields)));
+    return mergedSelectionSetBuilder.build();
```

Comment:
This build seems lossy as there is other data in selection set which are not copied over: 
`sourceLocation`, `comments` , `ignoredChars` and `additionalData`